src/mlpack/methods/cf/ncf_impl.hpp

```diff
@@ -0,0 +1,430 @@
+
+#include <mlpack/prereqs.hpp>
+#include "ncf.hpp"
+
+namespace mlpack {
+namespace cf {
+/**
+ * Construct the NCF object using the desired algorithm and optimizer.
+ */
+template<typename OptimizerType>
+NCF::NCF(arma::mat& dataset,
+         std::string algorithm,
+         OptimizerType& optimizer,
+         const size_t embedSize,
+         const size_t neg,
+         const size_t epochs,
+         bool implicit):
+    dataset(dataset),
+    neg(neg),
+    epochs(epochs),
+    embedSize(embedSize),
+    implicit(implicit)
+{
+  if (embedSize < 1)
+  {
+    Log::Warn << "NCF::NCF(): Embedding size should be > 0 ("
+        << embedSize << " given). Setting value to 8.\n";
+    // Set default value of 8.
+    this->embedSize = 8;
+  }
+  numUsers = (size_t) max(dataset.row(0)) + 1;
+  numItems = (size_t) max(dataset.row(1)) + 1;
+
+  if (algorithm == "GMF")
+  {
+    CreateGMF();
+  }
+  else if (algorithm == "MLP")
+  {
+    CreateMLP();
+  }
+  else if (algorithm == "NeuMF")
+  {
+    CreateNeuMF();
+  }
+
+  Train(optimizer);
+}
+
+/**
+ * Compute all unrated items for each user.
+ */
+void NCF::FindNegatives()
+{
+  negatives.clear();
+
+  for (size_t i = 0; i< numUsers; i++)
+  {
+    // Find items the user has rated.
+    arma::uvec userRates = arma::find(dataset.row(0) == i);
+    arma::mat itemRates = dataset.cols(userRates);
+
+    itemRates.shed_row(0);
+    itemRates.shed_row(1);
+    // List of all items.n
+    arma::vec negativeList = arma::linspace<arma::vec>(
+        0, numItems - 1, numItems);
+
+    for (size_t j = 0; j < itemRates.n_cols; j++)
+    {
+      // Remove items which have been rated.
+      arma::uvec temp = arma::find(negativeList ==  itemRates(j));
+      if(temp.n_rows != 0)
+        negativeList.shed_row(temp(0));
+    }
+
+    std::vector<double> negList;
+    // Add all negatives to a vector.
+    negList = arma::conv_to<std::vector<double>>::from(negativeList);
+    negatives.push_back(negList);
+  }
+}
+
+/**
+ * Create training instances using both positive and negative instances.
+ */
+void NCF::GetTrainingInstance(arma::mat& predictors,
+                              arma::mat& responses)
+{
+  size_t q = 0, temp;
+  arma::colvec users(dataset.n_cols * (neg+1));
+  arma::colvec resp(dataset.n_cols * (neg+1)), items(dataset.n_cols * (neg+1));
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    temp = neg;
+
+    // Rating exists.
+    users(q) = dataset(0, i) + 1;
+    items(q) = dataset(1, i) + 1;
+    resp(q) = implicit ? 1 : dataset(2, i);
+    q++;
+    // From find negatives.
+    size_t val = negatives[dataset(0, i)].size();
+
+    while (temp != 0)
+    {
+      size_t j = math::RandInt(val);
+      // Add negatives.
+      users(q) = dataset(0, i) + 1;
+      items(q) = negatives[dataset(0, i)][j] + 1;
+      resp(q) = 0;
+      q++;
+      temp--;
+    }
+  }
+  predictors = arma::join_vert(users, items);
+  responses = arma::conv_to<arma::mat>::from(resp);
```

Comment:
To get the optimizer the actual number of functions, we have to set: ``numFunctions = responses.n_cols;`` at the end. ``numFunctions`` is of type ``size_t``.

Reply:
The optimizer will have access the ``numFunctions`` parameter via the introduced ``NumFunctions()`` method; see comment above.